src/shared/fdset.c

```diff
@@ -133,18 +136,28 @@ bool fdset_contains(FDSet *s, int fd) {
                 return false;
         }
 
-        return !!set_get(MAKE_SET(s), FD_TO_PTR(fd));
+        return hashmap_contains(MAKE_HASHMAP(s), FD_TO_PTR(fd));
 }
 
 int fdset_remove(FDSet *s, int fd) {
         assert(s);
-        assert(fd >= 0);
 
         /* Avoid integer overflow in FD_TO_PTR() */
         if (fd == INT_MAX)
                 return log_debug_errno(SYNTHETIC_ERRNO(ENOENT), "Refusing invalid fd: %d", fd);
 
-        return set_remove(MAKE_SET(s), FD_TO_PTR(fd)) ? fd : -ENOENT;
+        void *raw = hashmap_remove(MAKE_HASHMAP(s), fd >= 0 ? FD_TO_PTR(fd) : INT_TO_PTR(fd));
```

Comment:
so we should take some care here regarding index zero. 

i.e. INT_TO_PTR() does a direct cast, i.e. index 0 would be mapped to NULL. But that's something we want to avoid, since hashmap_get() will return NULL also when an entry is missing, and thus the result would be ambigous. For FD_TO_PTR() we actually add `+1` internally, to make sure we can distinguish fd 0 from NULL. 

We should do the same here. Maybe add a local macro helper FD_INDEX_TO_PTR() PTR_TO_FD_INDEX() that do this check

Reply:
With the refactor to have separate fdset_remove as suggested index=0 can no longer happen, and it is checked - I've also added an assert to make sure